pkg/client/authentication.go

```diff
@@ -0,0 +1,50 @@
+package client
+
+import (
+	"context"
+	"fmt"
+	"net/http"
+)
+
+type basicAuthRoundTripper struct {
+	username string
+	password string
+	rt       http.RoundTripper
+}
+
+// RoundTrip is required to implement RoundTripper interface.
+// We check if an authorization token is already set, if not we set it.
+// We return the initial RoundTripper to chain it in the whole process.
+func (b *basicAuthRoundTripper) RoundTrip(req *http.Request) (*http.Response, error) {
+	if len(req.Header.Get("Authorization")) != 0 {
+		resp, err := b.rt.RoundTrip(req)
+		if err != nil {
+			return nil, fmt.Errorf("inner round trip error: %w", err)
+		}
+
+		return resp, nil
+	}
+
+	req = req.Clone(context.TODO())
```

Comment:
```suggestion
	req = req.Clone(req.Context())
```

Otherwise using round tripper swallows the context control on the request:
```go
package client_test

import (
  "context"
  "errors"
  "net/http"
  "testing"
  "time"

  "github.com/flatcar-linux/fleetlock/pkg/client"
)

func Test_Cancelling_context_for_request_performed_with_http_client_with_basic_auth_round_tripper_cancels_the_request(t *testing.T) {
  httpClient := http.Client{
    Transport: client.NewBasicAuthRoundTripper("foo", "bar", nil),
  }

  requestTimeout := time.Second

  ctx, cancel := context.WithTimeout(context.Background(), requestTimeout)
  t.Cleanup(cancel)

  req, err := http.NewRequestWithContext(ctx, "GET", "http://10.255.255.1", nil)
  if err != nil {
    t.Fatal(err)
  }

  errCh := make(chan error, 1)
  go func() {
    _, err = httpClient.Do(req)
    errCh <- err
  }()

  testDeadline := time.NewTimer(2 * requestTimeout)
  select {
  case <-testDeadline.C:
    t.Fatalf("Expected request to return before the deadline")
  case err := <-errCh:
    if err != nil && !errors.Is(err, context.DeadlineExceeded) {
      t.Fatal(err)
    }
  }
}
```